File: typescript-test-playwright/tests/api/blog-posts/add-update.test.ts
import { expect, test } from '@playwright/test';
import { BlogPosts } from '@service/blog-posts';
import { StatusCode } from '@constant/http-response-codes';
import { BlogPost, NewBlogPost, UpdateBlogPost } from '@type/blog-post';

test.describe('Add update blog posts', () => {
  let blogPosts: BlogPosts;
  let blogPost: BlogPost;

  test.beforeAll('Get a blog post to use for searching and updates', async ({ request }) => {
    blogPosts = new BlogPosts(request);

    await blogPosts.getAllPosts().then(async (response) => {
      expect(response.status()).toEqual(StatusCode.Ok);
      blogPost = (await response.json()).at(0);
    });
  });

  test.describe('Create blog post', () => {
    test('should add a new blog post', async () => {
      const newPost: NewBlogPost = {
        title: 'New Title',
        body: 'New blog details',
        userId: 1,
      };

      await blogPosts.addPost(newPost).then(async (response) => {
        expect(response.status()).toEqual(StatusCode.Created);
        expect(await response.json()).toHaveProperty('id');
      });
    });
  });

  test.describe('Update blog posts', () => {
    test('Should update an existing blog post', async () => {
      const postUpdate: UpdateBlogPost = {
        id: blogPost.id,
        title: 'this is the updated title',
      };

      await blogPosts.updatePost(postUpdate).then(async (response) => {
        expect(response.status()).toEqual(StatusCode.Ok);
        expect(await response.json()).toEqual(postUpdate);
      });
    });

    test('Should return an internal server error when updating a post that does not exist', async () => {
      const postUpdate: UpdateBlogPost = {
        // @ts-expect-error - Ignore type number requirement
        id: 'error',
        title: 'this should not update',
      };

      await blogPosts.updatePost(postUpdate).then((response) => {
        expect(response.status()).toEqual(StatusCode.InternalServerError);
      });
    });
  });
});

File: python-test/test/api/add_update_test.py
import pytest
from enums.http_enums import HttpStatusCode

from services.blog_posts.blog_posts_service import BlogPost, BlogPostsService, NewBlogPost, UpdateBlogPost
from services.utils.http_assertion_utils import HttpAssertion
from services.utils.json_utils import json_to_dictionary

@pytest.mark.blogpost
class TestExample():

    def setup(self):
        self.blog_post_service = BlogPostsService()
        self.expect = HttpAssertion()

        blog_posts_response = self.blog_post_service.get_all_posts()
        blog_posts_response.raise_for_status()
        self.blog_post: BlogPost = json_to_dictionary(blog_posts_response.text)[0]

    def test_should_add_a_new_blog_post(self):
        new_blog_post = NewBlogPost(userId=1, title='New Title', body='New blog details')

        blog_posts_response = self.blog_post_service.add_post(new_blog_post)
        self.expect(blog_posts_response.status_code).to_equal(HttpStatusCode.CREATED.value)

    def test_should_update_an_existing_blog_post(self):
        update_blog_post = UpdateBlogPost(id=self.blog_post.id, title='This is the updated title')

        blog_posts_response = self.blog_post_service.update_post(update_blog_post)
        self.expect(blog_posts_response.status_code).to_equal(HttpStatusCode.OK.value)

        blog_post: BlogPost = json_to_dictionary(blog_posts_response.text)
        self.expect(blog_post.title).to_equal(update_blog_post.title)

    def test_should_return_an_error_when_updating_a_post_that_does_not_exist(self):
        update_blog_post = UpdateBlogPost(id='error', title='This should not update')

        blog_posts_response = self.blog_post_service.update_post(update_blog_post)
        self.expect(blog_posts_response.status_code).to_equal(HttpStatusCode.INTERNAL_SERVER_ERROR.value)
